### DISCUSSION TITLE
Results of trying Terminal.Gui (and others) on GhostBSD (FreeBSD)

### DISCUSSION THREAD

**Discussion by sjsepan3** [tui-cs/Terminal.Gui#3529] [General]
# results of trying Terminal.Gui (and others) on GhostBSD (FreeBSD)

## Overview

Although FreeBSD it is not supported, since I am able to run a port of Code-OSS and also of .Net SDK 8.0 on GhostBSD 2024040.1 (FreeBSD 14.0-STABLE), I thought I'd try porting some of my .Net projects from Linux (Mint).

- A simple **console** app builds/runs OK
- A **Terminal.Gui** app builds OK, but gets runtime error "Curses failed to initialize, the exception is: Error loading native library "libncursesw.so.6, libncursesw.so.5", even after ncurses library 6.4_1 installed.
- An **Avalonia** app gets build warnings about obsolete (deprecated) dialog calls (as it does on Linux) but no errors (the Linux app runs OK) -- but running  on GhostBSD gives "No runtime platform services configured." I've looked at  "Cannot start Avalonia under FreeBSD #6990 ", and I am not referencing Avalonia.Native (but should I?)
- A **Modern.Forms** app build has no warnings or errors, but at runtime: "The type initializer for 'Modern.Forms.Form' threw an exception."
- A **Gtk.Sharp** app build has no warnings or errors, but at runtime: "The type initializer for 'Gtk.Application' threw an exception." even after installing the gtk-sharp30, gtk-sharp2.0, gio-sharp pkgs

I'm not really expecting a fix; just putting this here for future reference/discussion. The Terminal.Gui Linux app I used is this one [https://gitlab.com/sjsepan/TerminalGuiConsole1](https://gitlab.com/sjsepan/TerminalGuiConsole1)

## Code-OSS

```
Version: 1.88.1
Commit: e170252f762678dec6ca2cc69aba1570769a5d39
Date: 2024-05-07T13:49:21.988Z
Electron: 28.3.1
ElectronBuildId: undefined
Chromium: 120.0.6099.291
Node.js: 18.18.2
V8: 12.0.267.19-electron.0
OS: FreeBSD x64 14.0-STABLE
```

## dotnet --info

```
.NET SDK:
 Version:           8.0.105
 Commit:            eae90abaaf
 Workload version:  8.0.100-manifests.796a77f8

Runtime Environment:
 OS Name:     FreeBSD
 OS Version:  14
 OS Platform: FreeBSD
 RID:         freebsd-x64
 Base Path:   /usr/home/sjsepan/.dotnet/sdk/8.0.105/

.NET workloads installed:
 Workload version: 8.0.100-manifests.796a77f8
There are no installed workloads to display.

Host:
  Version:      8.0.5
  Architecture: x64
  Commit:       087e15321b

.NET SDKs installed:
  6.0.403 [/usr/home/sjsepan/.dotnet/sdk]
  8.0.105 [/usr/home/sjsepan/.dotnet/sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.App 6.0.11 [/usr/home/sjsepan/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.5 [/usr/home/sjsepan/.dotnet/shared/Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 6.0.11 [/usr/home/sjsepan/.dotnet/shared/Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.5 [/usr/home/sjsepan/.dotnet/shared/Microsoft.NETCore.App]

Other architectures found:
  None

Environment variables:
  DOTNET_ROOT       [/home/sjsepan/.dotnet]

global.json file:
  Not found

Learn more:
  https://aka.ms/dotnet/info

Download .NET:
  https://aka.ms/dotnet/download
```

**Comment by tznind** [comment]
That's really cool thanks, I think v2 (see v2_develop branch) is going to do away with curses dependency so that would resolve that particular issue.

You may also be able to get around this by using `Application.UseSystemConsole = true;`

**Reply by sjsepan3** [reply]
> That's really cool thanks, I think v2 (see v2_develop branch) is going to do away with curses dependency so that would resolve that particular issue.
> 
> You may also be able to get around this by using `Application.UseSystemConsole = true;`

Interesting! Thank You for that, and also for the tip -- I'll give that a try! :-)

UPDATE: Bingo! That was it @tznind ! I added that line before Init and it was good to go. Thanks!

![Screenshot from 2024-06-05 15-17-03](https://github.com/gui-cs/Terminal.Gui/assets/121126599/c58426b1-8e37-4f2a-8dec-0f6924949ce8)

**Reply by BDisp** [reply]
That's XP OS?

**Reply by sjsepan3** [reply]
> That's XP OS?

:-) @BDisp Nope, that's a screenshot of GhostBSD (Dell OptiPlex 7020) with the Chicago95 theme (ClassicLooks Bronze controls / WinMe window border / SE98 icons) as seen through a VNC session in Remmina,  from Linux Mint Cinnamon (Dell OptiPlex 5040) with my own custom theme ([Windows_XP_NT_Theme](https://gitlab.com/sjsepan/Windows_XP_NT_Theme)) (derived from one of the excellent Gtk themes 'Windows XP' by [Elbullazul ](https://github.com/B00merang-Project) on B00merang-Project).

**Comment by dodexahedron** [comment]
Very cool.

I've been curious how much effort it would take to enable additional platform support like `[a-zA-Z][a-zA-Z0-9]*BSD`, when .net is available on those platforms. Awesome that you A) tried it, and B) took the time to report your findings!

Thanks!

**Reply by sjsepan3** [reply]
> Very cool.
> 
> I've been curious how much effort it would take to enable additional platform support like `[a-zA-Z][a-zA-Z0-9]*BSD`, when .net is available on those platforms. Awesome that you A) tried it, and B) took the time to report your findings!
> 
> Thanks!

Thanks, and Welcome! To be fair, I'm not going to be exhaustively trying many  of the BSDs; I found it challenging to get FreeBSD installed. But GhostBSD was an easy install. ;-)

**Reply by dodexahedron** [reply]
Hey I'm floored you did what you did in the first place. No worries. 🥳

**Reply by BDisp** [reply]
@sjsepan3 please check if you have a file like `libncurses6*` in the `/usr/local/lib`. If you have then you'll need to add this local to the `PATH` env.

**Reply by sjsepan3** [reply]
There are these...I could add this location to the path  if it is not already ... (just checked, it does not appear to be yet).

![Screenshot from 2024-06-07 21-29-22](https://github.com/gui-cs/Terminal.Gui/assets/121126599/24ed8030-a318-4f3d-b214-c25100c4030f)

**Reply by BDisp** [reply]
Yes, on the ~/.profile add this line and I think the `CursesDriver` will work. But you need to comment on the code the `UseSystemConsole`.

`export PATH=$PATH:/usr/local/lib`

**Reply by sjsepan3** [reply]
Thanks -- I will be trying that today, then!

**Update:** its done, but maybe I missed something; same error.
-put it in .profile
-rebooted
-it is in PATH
-rebuilt without UseSystemConsole setting

![Screenshot from 2024-06-08 08-03-32](https://github.com/gui-cs/Terminal.Gui/assets/121126599/58e4f631-5a58-4ab1-a9f1-08808aeb1a90)

I wonder if I'm missing some ENV variable or link

![screenshot](https://github.com/gui-cs/Terminal.Gui/assets/121126599/7c1dfea4-d11d-4195-a8d8-94b2c35c4c78)

**Reply by BDisp** [reply]
I think GhostBSD doesn't use coreclr library and I think Linux uses this version of libncurses. Can you debug with coreclr in vscode?

**Reply by sjsepan3** [reply]
> I think GhostBSD doesn't use coreclr library and I think Linux uses this version of libncurses. Can you debug with coreclr in vscode?

I'll see if I can determine that... 

So this is what I get trying to debug in Code:
![Screenshot from 2024-06-10 12-27-00](https://github.com/gui-cs/Terminal.Gui/assets/121126599/aed92928-8bf0-4fb9-bb5e-a8b3fa593d00)

**Reply by dodexahedron** [reply]
Oh hey. About curses. Be aware we do have eliminating dependency on curses for unixy platforms as a goal for V2.

Do you by any chance have any thoughts about that, with regards to *BSD?

**Reply by sjsepan3** [reply]
@dodexahedron: I think @tznind mentioned something about that too, and I know of no reason to object. Thanks!

**Comment by tig** [comment]
Damn cool. Thanks for this!
